javaslang/src/main/java/javaslang/collection/Tree.java

```diff
@@ -334,11 +348,7 @@ default int leafCount() {
      * @return The number of nodes of this tree.
      */
     default int nodeCount() {
-        if (isEmpty()) {
-            return 0;
-        } else {
-            return 1 + getChildren().foldLeft(0, (count, child) -> count + child.nodeCount());
-        }
+        return length();
```

Comment:
This is a very good question! In 2013 (I think) I attended the Coursera course _progfun_ by Martin Odersky/EPFL. There was a assignment where we had to implement some data structure using traits (I think). In my first submission I missed some points because I didn't implemented a method similar to this default method.

I thought about it and now I design the interfaces/classes the way that only methods that are atomic / cannot be deduced by other methods are implemented within the implementing classes (like List.Nil, List.Cons). Other methods are implemented as default method within the interface. I think in general this makes it more easy to provide different implementations for one interface - all reduces to the most important methods.